flink-runtime/src/test/java/org/apache/flink/runtime/scheduler/declarative/WaitingForResourcesTest.java

```diff
@@ -0,0 +1,328 @@
+
+        assertThat(
+                ctx.getStateTransitions(),
+                contains(Tuple2.of(MockContext.States.EXECUTING, JobStatus.RUNNING)));
+    }
+
+    @Test
+    public void testTransitionToFinishedOnGlobalFailure() {
+        MockContext ctx = new MockContext();
+        ctx.setHasEnoughResources(() -> false);
+
+        WaitingForResources wfr = new WaitingForResources(ctx, log, ResourceCounter.empty());
+        wfr.onEnter();
+        wfr.handleGlobalFailure(new RuntimeException("mock failure"));
+        assertThat(
+                ctx.getStateTransitions(),
+                contains(Tuple2.of(MockContext.States.FINISHED, JobStatus.INITIALIZING)));
+    }
+
+    @Test
+    public void testCancel() {
+        MockContext ctx = new MockContext();
+        ctx.setHasEnoughResources(() -> false);
+
+        WaitingForResources wfr = new WaitingForResources(ctx, log, ResourceCounter.empty());
+        wfr.onEnter();
+        wfr.cancel();
+        assertThat(
+                ctx.getStateTransitions(),
+                contains(Tuple2.of(MockContext.States.FINISHED, JobStatus.CANCELED)));
+    }
+
+    @Test
+    public void testSuspend() {
+        MockContext ctx = new MockContext();
+        ctx.setHasEnoughResources(() -> false);
+
+        WaitingForResources wfr = new WaitingForResources(ctx, log, ResourceCounter.empty());
+        wfr.onEnter();
+        wfr.suspend(new Exception("test"));
+        assertThat(
+                ctx.getStateTransitions(),
+                contains(Tuple2.of(MockContext.States.FINISHED, JobStatus.SUSPENDED)));
+    } */
+
+    private class StateTestContext {
+        private final MockContext mockContext;
+        private WaitingForResources waitingForResources;
+
+        public StateTestContext(Consumer<MockContext> modifyContext) {
+            this.mockContext = new MockContext();
+            modifyContext.accept(mockContext);
+            this.waitingForResources =
+                    new WaitingForResources(mockContext, log, ResourceCounter.empty());
+            this.mockContext.setWaitingForResources(waitingForResources);
+            waitingForResources.onEnter();
+        }
+
+        public MockContext getMockContext() {
+            return mockContext;
+        }
+
+        public void assertStateTransitionTo(Class<? extends State> expected) {
+            Assert.assertThat(mockContext.getTransitionTo(), instanceOf(expected));
+        }
+
+        public void assertNoStateTransition() {
+            Assert.assertThat(mockContext.getTransitionTo(), nullValue());
+        }
+
+        public WaitingForResources getWaitingForResources() {
+            return waitingForResources;
+        }
+    }
+
+    private static class MockContext implements WaitingForResources.Context {
+
+        private static final Logger LOG = LoggerFactory.getLogger(MockContext.class);
+
+        private WaitingForResources waitingForResources;
+
+        private State transitionTo = null;
```

Comment:
I would suggest to not store the last state here. This effectively couples all states together.